File: packages/typescript/src/scorers/rouge.ts
import { EvalCase, EvalResult } from "../models";
import { Scorer, makeResult } from "./base";

function lcsLength(a: string[], b: string[]): number {
  const m = a.length, n = b.length;
  const dp: number[][] = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      dp[i][j] = a[i - 1] === b[j - 1] ? dp[i - 1][j - 1] + 1 : Math.max(dp[i - 1][j], dp[i][j - 1]);
    }
  }
  return dp[m][n];
}

export function rougeL(hypothesis: string, reference: string): number {
  const h = hypothesis.toLowerCase().split(/\s+/).filter(Boolean);
  const r = reference.toLowerCase().split(/\s+/).filter(Boolean);
  if (h.length === 0 || r.length === 0) return 0;
  const lcs       = lcsLength(h, r);
  const precision = lcs / h.length;
  const recall    = lcs / r.length;
  if (precision + recall === 0) return 0;
  return (2 * precision * recall) / (precision + recall);
}

export class RougeScorer implements Scorer {
  readonly name = "rouge_l";

  constructor(private threshold: number = 0.5) {}

  score(evalCase: EvalCase, output: string): EvalResult {
    const s = rougeL(output, evalCase.expected ?? "");
    return makeResult(evalCase, output, s, this.threshold);
  }
}

File: packages/python/evalforge/scorers/rouge.py
from ..models import EvalCase, EvalResult
from .base import Scorer


def _lcs_length(a: list, b: list) -> int:
    m, n = len(a), len(b)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    return dp[m][n]


def rouge_l(hypothesis: str, reference: str) -> float:
    h = hypothesis.lower().split()
    r = reference.lower().split()
    if not h or not r:
        return 0.0
    lcs = _lcs_length(h, r)
    precision = lcs / len(h)
    recall    = lcs / len(r)
    if precision + recall == 0:
        return 0.0
    return 2 * precision * recall / (precision + recall)


class RougeScorer(Scorer):
    name = "rouge_l"

    def __init__(self, threshold: float = 0.5):
        self.threshold = threshold

    def score(self, case: EvalCase, output: str) -> EvalResult:
        s = rouge_l(output, case.expected or "")
        return self._result(case, output, s, pass_threshold=self.threshold)
